src/vsr/replica.zig

```diff
@@ -6700,152 +6717,186 @@ const DVCQuorum = struct {
+
         const dvcs = DVCQuorum.dvcs_canonical(dvc_quorum);
+        assert(dvcs.len > 0);
+        assert(dvcs.len <= dvcs_total);
 
-        const op_head_max = op_max_canonical(dvc_quorum);
-        // The number of uncommitted ops cannot be more than the length of the pipeline.
-        const op_suffix_min = op_head_max -| constants.pipeline_prepare_queue_max;
-        assert(op_suffix_min <= op_head_max);
+        const op_head_max = DVCQuorum.op_max_canonical(dvc_quorum);
+        const op_head_min = DVCQuorum.commit_max(dvc_quorum);
 
-        var op_head_min = op_suffix_min;
-        var ops_in_suffix = std.StaticBitSet(constants.pipeline_prepare_queue_max).initEmpty();
-        for (dvcs.constSlice()) |message| {
-            const message_headers = message_body_as_headers_chain_disjoint(message);
-            for (message_headers) |header| {
-                if (header.op > op_suffix_min) {
-                    ops_in_suffix.set((header.op - op_suffix_min) - 1);
+        // Iterate the highest definitely committed op and all maybe-uncommitted ops.
+        var op = op_head_min;
+        const op_head = while (op <= op_head_max) : (op += 1) {
+            // Replicas with non-canonical log-views implicitly nack all uncommitted ops.
+            var nacks: usize = dvcs_total - dvcs.len;
+
+            const found = for (dvcs.constSlice()) |dvc| {
+                if (dvc.header.op < op) {
+                    nacks += 1;
+                    continue;
                 }
-            }
-            op_head_min = std.math.max(op_head_min, message_headers[message_headers.len - 1].op);
-        }
-        assert(op_head_max == 0 or ops_in_suffix.isSet((op_head_max - op_suffix_min) - 1));
-        assert(op_head_min >= op_suffix_min);
-        assert(op_head_min <= op_head_max);
 
-        const op_head = blk: {
-            var op = op_head_min + 1;
-            while (op < op_head_max) : (op += 1) {
-                if (!ops_in_suffix.isSet((op - op_suffix_min) - 1)) {
-                    break :blk op - 1;
+                const headers = message_body_as_view_headers(dvc);
+                const header_index = dvc.header.op - op;
+                if (header_index >= headers.slice.len) {
+                    // This op is definitely committed by the cluster.
+                } else {
+                    const header = &headers.slice[header_index];
+                    assert(header.op == op);
+
+                    switch (vsr.Headers.dvc_header_type(header)) {
+                        .blank => nacks += 1,
+                        .fault => {},
+                        .valid => break true,
+                    }
                 }
+            } else false;
+
+            if (found) {
+                // This op is eligible to be the view's head.
             } else {
-                break :blk op_head_max;
+                // Never nack op_head_min (aka commit_max).
```

Comment:
Shouldn't this be an assert? The op is committed, so it can't be nacked, right?

Reply:
Good point! There may be a few nacks, but never a quorum.